Source/WTF/wtf/ThreadSafeWeakPtr.h

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Lock.h>
+#include <wtf/MainThread.h>
+#include <wtf/RefPtr.h>
+
+namespace WTF {
```

Comment:
I think it'd be worth having a little explanation of the class at the top here, particularly to point out that this is the only safe way to have thread-safe refcounting+weakptrs, but does come at the cost of mutex locking for every ref/deref (strong and weak), and some consumers might be better off looking for patterns that don't need this combination.

Reply:
I'm planning to add some static_asserts to WeakPtr stuff to make it a compiler error to have a WeakPtr to a ThreadSafeRefCounted type, but I can't add those until the transition completes.  I don't think such a comment would add much.

Reply:
I think the performance overhead of locking is a bit unexpected (since the other refcounting variants don't have that), and consumers might want to reconsider whether they really need WeakPtrs. Explaining that trade-off upfront seems useful to me :)

Big fan of the static_asserts to block the existing (broken) pattern of ThreadSafeRefCounted+CanMakeWeakPtr!

Reply:
WTF::Lock is pretty fast, and most of the time it'll just be an atomic operation and a branch.  Even with lots of concurrent threads it'll mostly just be a few spins and you have the lock.  If someone is reconsidering whether they really need ThreadSafeWeakPtr, we don't want them to start storing raw pointers.  I really don't think adding a comment saying "Watch out, this uses locks" will be useful at all.

Reply:
After taking Fujii's excellent suggestion, there will never be more than ~10 operations before releasing the lock, so even in the worst case it will only take a few spins to get the lock.